apps/omg_child_chain/lib/omg_child_chain/fee_server.ex

```diff
@@ -39,78 +41,141 @@ defmodule OMG.ChildChain.FeeServer do
+  def accepted_fees() do
+    {:ok, load_accepted_fees()}
+  end
+
+  @doc """
+  Returns currently accepted tokens and amounts in which transaction fees are collected for each transaction type
+  """
+  @spec current_fees() :: {:ok, Fees.full_fee_t()}
+  def current_fees() do
+    {:ok, load_current_fees()}
+  end
+
+  def handle_info(:expire_previous_fees, state) do
+    true = :ets.update_element(:fees_bucket, :previous_fees, {2, nil})
+    _ = Logger.info("Previous fees are now invalid and current fees must be paid")
+    true = save_merged_fees()
+    {:noreply, state}
   end
 
   def handle_info(:update_fee_specs, state) do
-    _ =
-      case update_fee_specs(state[:fee_adapter]) do
-        :ok ->
+    new_state =
+      case update_fee_specs(state) do
+        {:ok, updated_state} ->
           Alarm.clear(Alarm.invalid_fee_file(__MODULE__))
+          updated_state
 
-        _ ->
+        {:error, _, updated_state} ->
           Alarm.set(Alarm.invalid_fee_file(__MODULE__))
+          updated_state
       end
 
-    {:noreply, state}
+    {:noreply, new_state}
   end
 
-  @spec update_fee_specs(module()) :: :ok | {:error, atom() | [{:error, atom()}, ...]}
-  defp update_fee_specs(adapter) do
+  @spec update_fee_specs(map()) :: :ok | {:error, atom() | [{:error, atom()}, ...]}
+  defp update_fee_specs(%{fee_adapter: adapter, expire_fee_timer: current_timer} = state) do
     source_updated_at = :ets.lookup_element(:fees_bucket, :fee_specs_source_updated_at, 2)
 
     case adapter.get_fee_specs(source_updated_at) do
       {:ok, fee_specs, source_updated_at} ->
         :ok = save_fees(fee_specs, source_updated_at)
-        _ = Logger.info("Reloaded #{inspect(Enum.count(fee_specs))} fee specs from
-                         #{inspect(adapter)}, changed at #{inspect(source_updated_at)}")
-        :ok
+        _ = Logger.info("Reloaded fee specs from #{inspect(adapter)}, changed at #{inspect(source_updated_at)}")
+        new_timer = start_expiration_timer(current_timer)
+
+        _ =
+          Logger.info(
+            "Timer started: previous fees will still be valid for #{inspect(@fee_buffer_duration_ms)} ms, or until new fees are set"
+          )
+
+        {:ok, Map.put(state, :expire_fee_timer, new_timer)}
 
       :ok ->
-        :ok
+        {:ok, state}
 
       error ->
         _ = Logger.error("Unable to update fees from file. Reason: #{inspect(error)}")
-        error
+        {:error, error, state}
     end
   end
 
-  defp save_fees(fee_specs, last_updated_at) do
+  defp save_fees(new_fee_specs, last_updated_at) do
+    current_fee_specs = :ets.lookup_element(:fees_bucket, :fees, 2)
+
     true =
       :ets.insert(:fees_bucket, [
         {:updated_at, :os.system_time(:second)},
         {:fee_specs_source_updated_at, last_updated_at},
-        {:fees, fee_specs}
+        {:fees, new_fee_specs},
+        {:previous_fees, current_fee_specs}
       ])
 
+    true = save_merged_fees()
```

Comment:
since you're doing two separate writes with related content, I'll quickly validate your assumptions.
are you OK if the process owning the ETS table dies? do you have a tests for the correct behaviour - if the process exits between `fees_bucket ` write and `fees_bucket` with key `merged_fees ` write?

Reply:
That's a good point, ideally we don't want to have a difference between these 2 values, I'll refactor a bit and write the `merged_fees` at the same time as the rest